subsys/net/l2/ethernet/gptp/gptp_mi.c

```diff
@@ -756,85 +771,60 @@ static void gptp_update_local_port_clock(void)
+	sync_offset_ts = sync_receipt_ts - sync_receipt_local_ts;
+	servo_ds->master_offset_ns = sync_offset_ts;
 
 	clk = net_eth_get_ptp_clock(GPTP_PORT_IFACE(port));
 	if (!clk) {
 		return;
 	}
 
-	if (second_diff > 0 && nanosecond_diff < 0) {
-		second_diff--;
-		nanosecond_diff = NSEC_PER_SEC + nanosecond_diff;
-	}
-
-	if (second_diff < 0 && nanosecond_diff > 0) {
-		second_diff++;
-		nanosecond_diff = -(int64_t)NSEC_PER_SEC + nanosecond_diff;
-	}
-
-	ptp_clock_rate_adjust(clk, port_ds->neighbor_rate_ratio);
-
 	/* If time difference is too high, set the clock value.
 	 * Otherwise, adjust it.
 	 */
-	if (second_diff || (second_diff == 0 &&
-			    (nanosecond_diff < -5000 ||
-			     nanosecond_diff > 5000))) {
-		bool underflow = false;
+	if (sync_offset_ts < -servo_ds->step_threshold_ns ||
+	    sync_offset_ts > servo_ds->step_threshold_ns) {
+		uint64_t current_ts;
 
 		key = irq_lock();
+		ptp_clock_adjust(clk, 0);
 		ptp_clock_get(clk, &tm);
 
-		if (second_diff < 0 && tm.second < -second_diff) {
-			NET_DBG("Do not set local clock because %lu < %ld",
-				(unsigned long int)tm.second,
-				(long int)-second_diff);
-			goto skip_clock_set;
-		}
-
-		tm.second += second_diff;
-
-		if (nanosecond_diff < 0 &&
-		    tm.nanosecond < -nanosecond_diff) {
-			underflow = true;
-		}
-
-		tm.nanosecond += nanosecond_diff;
+		current_ts = tm.second * NSEC_PER_SEC + tm.nanosecond;
+		current_ts += sync_offset_ts;
 
-		if (underflow) {
-			tm.second--;
-			tm.nanosecond += NSEC_PER_SEC;
-		} else if (tm.nanosecond >= NSEC_PER_SEC) {
-			tm.second++;
-			tm.nanosecond -= NSEC_PER_SEC;
-		}
+		tm.second = current_ts / NSEC_PER_SEC;
+		tm.nanosecond = current_ts % NSEC_PER_SEC;
 
-		/* This prints too much data normally but can be enabled to see
-		 * what time we are setting to the local clock.
-		 */
-		if (0) {
-			NET_INFO("Set local clock %lu.%lu",
-				 (unsigned long int)tm.second,
-				 (unsigned long int)tm.nanosecond);
+		/* Must tell user master clock is unstable. */
+		if (sync_offset_ts < 0) {
+			NET_ERR("Clock jump forward to %llu.%09u, offset is %lldns", tm.second,
+				tm.nanosecond, sync_offset_ts);
+		} else {
+			NET_WARN("Clock jump to %llu.%09u", tm.second, tm.nanosecond);
 		}
 
 		ptp_clock_set(clk, &tm);
-
-	skip_clock_set:
 		irq_unlock(key);
 	} else {
-		if (nanosecond_diff < -200) {
-			nanosecond_diff = -200;
-		} else if (nanosecond_diff > 200) {
-			nanosecond_diff = 200;
+		ki_err = sync_offset_ts * servo_ds->ki;
+		adjfreq = sync_offset_ts * servo_ds->kp + servo_ds->observed_drift + ki_err;
+
+		if (adjfreq > servo_ds->max_freq_adjust) {
+			adjfreq = servo_ds->max_freq_adjust;
+		} else if (adjfreq < -servo_ds->max_freq_adjust) {
+			adjfreq = -servo_ds->max_freq_adjust;
+		} else {
+			servo_ds->observed_drift += ki_err;
 		}
 
-		ptp_clock_adjust(clk, nanosecond_diff);
+		NET_DBG("offset from master: %lldns, adjfreq: %d",
+			servo_ds->master_offset_ns, (int)adjfreq);
+
+		ptp_clock_adjust(clk, (int)adjfreq);
```

Comment:
You are using here adjust and feed it with  a value called adjfreq. As I read the API, the adjust call makes a change in time = phase, rate_adjust makes a change in frequency.
I think the variable should be called time_adjust, then the Ki takes care of frequency correction, the Kp part of time / phase compensation.

Reply:
Thanks for your suggestion, using time_adjust can indeed describe this variable well.
It is good to specify time reference here, and the hardware driver of the manufacturer needs to implement the correction ratio of time and frequency.